File: biweekly/104/q3.cc
class Solution {
    using i64 = long long;
public:
    long long maximumOr(vector<int>& nums, int k) {
        int n = nums.size();
        std::vector<i64> pf(n + 1), sf(n + 1);
        for (int i = 0; i < n; ++i) {
            pf[i + 1] = (1LL * pf[i]) | nums[i];
        }
        for (int i = n; i > 0; --i) {
            sf[i - 1] = (1LL * sf[i]) | nums[i - 1];
        }
        
        i64 ans = 0;
        for (int i = 0; i < n; ++i) {
            ans = max(ans, pf[i] | ((1LL * nums[i]) << k) | sf[i + 1]);
        }
        return ans;
    }
};

File: weekly/349/go_a/q3.go
import (
    "math"
)

func minCost(nums []int, x int) int64 {
    n := len(nums)
    vs := make([]int, n)
    for i := range nums {
        vs[i] = nums[i]
    }
    ans := math.MaxInt
    for k := range nums {
        a := 0
        for i := range vs {
            vs[i] = min(vs[i], nums[(i + k) % n])
            a += vs[i]
        }
        a += x * k
        ans = min(ans, a)
    }
    return int64(ans)
}

func min(a, b int) int {
    if a < b {
        return a
    } else {
        return b
    }
}

func max(a, b int) int {
    return a + b - min(a, b)
}

